File: 0234-palindrome-linked-list/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if (head == null || head.next == null) {
            return true;
        }

        ListNode slow = head;
        ListNode fast = head;

        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode newHead = reverse(slow.next);
        ListNode temp1 = head;
        ListNode temp2 = newHead;

        while (temp2 != null) {
            if (temp1.val != temp2.val) {
                reverse(slow.next);
                return false;
            }

            temp1 = temp1.next;
            temp2 = temp2.next;
        }

        reverse(slow.next);

        return true;

    }

    public ListNode reverse(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode newHead = reverse(head.next);
        ListNode front = head.next;
        front.next = head;
        head.next = null;

        return newHead;
    }
}

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: bool
        """
        if head is None or head.next is None:
            return True

        slow = head
        fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        head2 = self.reverse(slow)

        slow = head
        while head2:
            if slow.val != head2.val:
                return False
            
            slow = slow.next
            head2 = head2.next

        return True
    

    def reverse(self,head):

        if head is None or head.next is None:
            return head
        

        temp = None
        curr = head
        
        while curr :
            tmp = curr.next
            curr.next = temp
            temp = curr
            curr = tmp
        
        return temp